packages/valory/skills/termination_abci/rounds.py

```diff
@@ -0,0 +1,126 @@
+    TransactionSubmissionAbciApp,
+)
+
+
+class Event(Enum):
+    """Defines the (background) round events."""
+
+    TERMINATE = "terminate"
+
+
+class SynchronizedData(BaseSynchronizedData):
+    """
+    Class to represent the synchronized data.
+
+    This data is replicated by the tendermint application.
+    """
+
+    @property
+    def safe_contract_address(self) -> Optional[str]:
+        """Get the safe contract address."""
+        return cast(Optional[str], self.db.get("safe_contract_address", None))
+
+    @property
+    def termination_majority_reached(self) -> bool:
+        """Get termination_majority_reached."""
+        return cast(bool, self.db.get("termination_majority_reached", False))
+
+
+class BackgroundRound(CollectSameUntilThresholdRound):
+    """Defines the background round, which runs concurrently with other rounds."""
+
+    round_id: str = "background_round"
+    allowed_tx_type = BackgroundPayload.transaction_type
+    payload_attribute: str = "background_data"
+    synchronized_data_class = SynchronizedData
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Event]]:
+        """Process the end of the block."""
+        if self.threshold_reached:
+            state = self.synchronized_data.update(
+                synchronized_data_class=self.synchronized_data_class,
+                termination_majority_reached=True,
+                most_voted_tx_hash=self.most_voted_payload,
+            )
+            return state, Event.TERMINATE
+
+        return None
+
+
+class TerminationRound(AbstractRound):
+    """Round to act as the counterpart of the behaviour responsible for terminating the agent."""
+
+    round_id = "termination_round"
+
+    def check_payload(self, payload: BaseTxPayload) -> None:
+        """No logic required here."""
+        pass
+
+    def process_payload(self, payload: BaseTxPayload) -> None:
+        """No logic required here."""
+        pass
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Enum]]:
+        """No logic required here."""
+        return None
+
+
+class PostTerminationTxAbciApp(AbciApp):
+    """The abci app running after the multisig transaction has been settled."""
+
+    initial_round_cls = TerminationRound
+    # the following is not needed, it is added to satisfy the round check
+    # the TerminationRound when run it terminates the agent, so nothing can come after it
+    transition_function = {TerminationRound: {Event.TERMINATE: TerminationRound}}
+
+
+termination_transition_function = {
+    FinishedTransactionSubmissionRound: PostTerminationTxAbciApp.initial_round_cls,
+    FailedRound: TransactionSubmissionAbciApp.initial_round_cls,
+}
+TerminationAbciApp = chain(
+    (
+        TransactionSubmissionAbciApp,
+        PostTerminationTxAbciApp,
+    ),
+    termination_transition_function,
+)
+
+TerminationAbciApp.transition_function[BackgroundRound] = {
+    Event.TERMINATE: TransactionSubmissionAbciApp.initial_round_cls,
+}
```

Comment:
Why is this being updated after ?

Reply:
To avoid having another 1 class abci app. On second thought, we can get rid of this and handle it in the AbciApp side.

Reply:
I think it's much more transparent if we have a number of single class apps. Then we chain them, then we append them via the new `add_termination`